sql/rls: retain function backrefs when updating policy dependencies

spilchen opened the pull request:
When a row-level security (RLS) policy references a user-defined function, we correctly establish both forward and backward references during creation. However, if a dependency of that function changes (e.g., a column used in the function's default expression is dropped), the schema changer updates the function’s dependencies but inadvertently drops the backreference to the RLS policy.

This leads to a missing backref, causing the schema changes to fail due. This change ensures the policy backref is preserved.

Fixes #147471

Release note (bug fix): Fixed a bug where functions lost their RLS policy backreferences, leading to schema change failures.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147696)
<!-- Reviewable:end -->

rafiss reviewed (approved):
nice find! just had one idea about testing

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147696#-:-ORwXrNkBNcFKmr5_d1O:b-oseulb)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 720 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147696#-ORwXl3N5ebnW6f4RoIO:-ORwXl3N5ebnW6f4RoIP:b28x540) ([raw file](https://github.com/cockroachdb/cockroach/blob/c52bfad07021932de4cf67b4bcef910a36ea942e/pkg/sql/logictest/testdata/logic_test/row_level_security#L720)):*
> ```
> CREATE POLICY p ON t USING (f());
> 
> # Function reference should be updated but still include dependency on policy.
> ```

could we do something to explicitly check that the dependency is still there? for example, in another test we do this: https://github.com/cockroachdb/cockroach/blob/9626614c9bb213c204ec76e794ea15a65d82461e/pkg/sql/logictest/testdata/logic_test/udf_in_column_defaults#L41


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/rls: retain function backrefs when updating policy dependencies

When a row-level security (RLS) policy references a user-defined
function, we correctly establish both forward and backward references
during creation. However, if a dependency of that function changes
(e.g., a column used in the function's default expression is dropped),
the schema changer updates the function’s dependencies but inadvertently
drops the backreference to the RLS policy.

This leads to a missing backref, causing the schema changes to fail due.
This change ensures the policy backref is preserved.

Fixes #147471

Release note (bug fix): Fixed a bug where functions lost their RLS
policy backreferences, leading to schema change failures.

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147696)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 720 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147696#-ORwXl3N5ebnW6f4RoIO:-OS3xKeT5JUzgPeFOGjH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c52bfad07021932de4cf67b4bcef910a36ea942e/pkg/sql/logictest/testdata/logic_test/row_level_security#L720)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could we do something to explicitly check that the dependency is still there? for example, in another test we do this: https://github.com/cockroachdb/cockroach/blob/9626614c9bb213c204ec76e794ea15a65d82461e/pkg/sql/logictest/testdata/logic_test/udf_in_column_defaults#L41
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
